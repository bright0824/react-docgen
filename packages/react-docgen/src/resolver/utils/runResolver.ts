import type FileState from '../../FileState.js';
import type { Resolver, ResolverClass, ResolverFunction } from '../index.js';

function isResolverClass(resolver: Resolver): resolver is ResolverClass {
  return typeof resolver === 'object';
}

export default function runResolver(
  resolver: Resolver,
  file: FileState,
): ReturnType<ResolverFunction> {
  return isResolverClass(resolver) ? resolver.resolve(file) : resolver(file);
}
