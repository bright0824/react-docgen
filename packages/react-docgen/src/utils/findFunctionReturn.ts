import type { NodePath } from '@babel/traverse';
import { visitors } from '@babel/traverse';
import resolveToValue from './resolveToValue';
import { ignore } from './traverse';

type Predicate = (p: NodePath) => boolean;
interface TraverseState {
  readonly predicate: Predicate;
  resolvedReturnPath?: NodePath;
  readonly seen: WeakSet<NodePath>;
}

const explodedVisitors = visitors.explode<TraverseState>({
  Function: { enter: ignore },
  Class: { enter: ignore },
  ObjectExpression: { enter: ignore },
  ReturnStatement: {
    enter: function (path, state) {
      const argument = path.get('argument');

      if (argument.hasNode()) {
        const resolvedPath = resolvesToFinalValue(
          argument,
          state.predicate,
          state.seen,
        );

        if (resolvedPath) {
          state.resolvedReturnPath = resolvedPath;
          path.stop();
        }
      }
    },
  },
});

function resolvesToFinalValue(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  // avoid returns with recursive function calls
  if (seen.has(path)) {
    return;
  }
  seen.add(path);

  // Is the path already passes then return it.
  if (predicate(path)) {
    return path;
  }

  const resolvedPath = resolveToValue(path);

  // If the resolved path is already passing then no need to further check
  // Only do this if the resolvedPath actually resolved something as otherwise we did this check already
  if (resolvedPath.node !== path.node && predicate(resolvedPath)) {
    return resolvedPath;
  }

  // If the path points to a conditional expression, then we need to look only at
  // the two possible paths
  if (resolvedPath.isConditionalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('consequent'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('alternate'), predicate, seen)
    );
  }

  // If the path points to a logical expression (AND, OR, ...), then we need to look only at
  // the two possible paths
  if (resolvedPath.isLogicalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('left'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('right'), predicate, seen)
    );
  }

  // If we have a call expression, lets try to follow it
  if (resolvedPath.isCallExpression()) {
    const returnValue = findFunctionReturnWithCache(
      resolveToValue(resolvedPath.get('callee')),
      predicate,
      seen,
    );

    if (returnValue) {
      return returnValue;
    }
  }

  return;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
function findFunctionReturnWithCache(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  let functionPath: NodePath = path;

  if (functionPath.isObjectProperty()) {
    functionPath = functionPath.get('value');
  } else if (functionPath.isClassProperty()) {
    const classPropertyValue = functionPath.get('value');

    if (classPropertyValue.hasNode()) {
      functionPath = classPropertyValue;
    }
  }

  if (!functionPath.isFunction()) {
    return;
  }

  // skip traversing for ArrowFunctionExpressions with no block
  if (path.isArrowFunctionExpression()) {
    const body = path.get('body');

    if (!body.isBlockStatement()) {
      return resolvesToFinalValue(body, predicate, seen);
    }
  }

  const state: TraverseState = {
    predicate,
    seen,
  };

  path.traverse(explodedVisitors, state);

  return state.resolvedReturnPath;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
export default function findFunctionReturn(
  path: NodePath,
  predicate: Predicate,
): NodePath | undefined {
  return findFunctionReturnWithCache(path, predicate, new WeakSet());
}
