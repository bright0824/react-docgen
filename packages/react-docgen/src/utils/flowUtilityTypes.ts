import type { NodePath } from '@babel/traverse';
import type { GenericTypeAnnotation } from '@babel/types';

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(
  path: NodePath,
): path is NodePath<GenericTypeAnnotation> {
  if (path.isGenericTypeAnnotation()) {
    const idPath = path.get('id');

    if (idPath.isIdentifier()) {
      const name = idPath.node.name;

      return name === '$Exact' || name === '$ReadOnly';
    }
  }

  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  let resultPath: NodePath = path;
  while (isSupportedUtilityType(resultPath)) {
    const typeParameters = resultPath.get('typeParameters');
    if (!typeParameters.hasNode()) break;

    const firstParam = typeParameters.get('params')[0];
    if (!firstParam) break;

    resultPath = firstParam;
  }

  return resultPath;
}
