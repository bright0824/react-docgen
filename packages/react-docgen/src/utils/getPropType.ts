import type { NodePath } from '@babel/traverse';
import { getDocblock } from '../utils/docblock.js';
import getMembers from './getMembers.js';
import getPropertyName from './getPropertyName.js';
import isRequiredPropType from '../utils/isRequiredPropType.js';
import printValue from './printValue.js';
import resolveToValue from './resolveToValue.js';
import resolveObjectKeysToArray from './resolveObjectKeysToArray.js';
import resolveObjectValuesToArray from './resolveObjectValuesToArray.js';
import type { PropTypeDescriptor } from '../Documentation.js';
import type { ArrayExpression, Expression, SpreadElement } from '@babel/types';

function getEnumValuesFromArrayExpression(
  path: NodePath<ArrayExpression>,
): Array<Record<string, unknown>> {
  const values: Array<Record<string, unknown>> = [];

  path.get('elements').forEach((elementPath) => {
    if (!elementPath.hasNode()) return;

    if (elementPath.isSpreadElement()) {
      const value = resolveToValue(elementPath.get('argument'));

      if (value.isArrayExpression()) {
        // if the SpreadElement resolved to an Array, add all their elements too
        return values.push(...getEnumValuesFromArrayExpression(value));
      } else {
        // otherwise we'll just print the SpreadElement itself
        return values.push({
          value: printValue(elementPath),
          computed: !elementPath.isLiteral(),
        });
      }
    }

    // try to resolve the array element to it's value
    const value = resolveToValue(elementPath as NodePath<Expression>);

    return values.push({
      value: printValue(value.isImportDeclaration() ? elementPath : value),
      computed: !value.isLiteral(),
    });
  });

  return values;
}

function getPropTypeOneOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const value = resolveToValue(argumentPath);

  if (value.isArrayExpression()) {
    type.value = getEnumValuesFromArrayExpression(value);
  } else {
    const objectValues =
      resolveObjectKeysToArray(value) || resolveObjectValuesToArray(value);

    if (objectValues) {
      type.value = objectValues.map((objectValue) => ({
        value: objectValue,
        computed: false,
      }));
    } else {
      // could not easily resolve to an Array, let's print the original value
      type.computed = true;
      type.value = printValue(argumentPath);
    }
  }

  return type;
}

function getPropTypeOneOfType(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  if (argumentPath.isArrayExpression()) {
    type.value = argumentPath.get('elements').map((elementPath) => {
      if (!elementPath.hasNode()) return;
      const descriptor: PropTypeDescriptor = getPropType(elementPath);
      const docs = getDocblock(
        elementPath as NodePath<Expression | SpreadElement>,
      );

      if (docs) {
        descriptor.description = docs;
      }

      return descriptor;
    });
  }

  return type;
}

function getPropTypeArrayOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const docs = getDocblock(argumentPath);

  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath);

  // @ts-ignore
  if (subType.name !== 'unknown') {
    type.value = subType;
  }

  return type;
}

function getPropTypeObjectOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const docs = getDocblock(argumentPath);

  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath);

  // @ts-ignore
  if (subType.name !== 'unknown') {
    type.value = subType;
  }

  return type;
}

function getFirstArgument(path: NodePath): NodePath | undefined {
  let argument: NodePath | undefined;

  if (path.isCallExpression()) {
    argument = path.get('arguments')[0];
  } else {
    const members = getMembers(path, true);

    if (members[0] && members[0].argumentPaths[0]) {
      argument = members[0].argumentPaths[0];
    }
  }

  return argument;
}

function isCyclicReference(
  argument: NodePath,
  argumentPath: NodePath,
): boolean {
  return Boolean(argument && resolveToValue(argument) === argumentPath);
}

/**
 * Handles shape and exact prop types
 */
function getPropTypeShapish(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  if (!argumentPath.isObjectExpression()) {
    argumentPath = resolveToValue(argumentPath);
  }

  if (argumentPath.isObjectExpression()) {
    const value: Record<string, PropTypeDescriptor> = {};
    let rawValue: string | undefined;

    argumentPath.get('properties').forEach((propertyPath) => {
      // We only handle ObjectProperty as there is nothing to handle for
      // SpreadElements and ObjectMethods
      if (propertyPath.isObjectProperty()) {
        const propertyName = getPropertyName(propertyPath);

        if (!propertyName) return;

        const valuePath = propertyPath.get('value');
        const argument = getFirstArgument(valuePath);

        // This indicates we have a cyclic reference in the shape
        // In this case we simply print the argument to shape and bail
        if (argument && isCyclicReference(argument, argumentPath)) {
          rawValue = printValue(argument);

          return;
        }

        const descriptor = getPropType(valuePath);
        const docs = getDocblock(propertyPath);

        if (docs) {
          descriptor.description = docs;
        }
        descriptor.required = isRequiredPropType(valuePath);
        value[propertyName] = descriptor;
      }
    });

    type.value = rawValue ?? value;
  }

  return type;
}

function getPropTypeInstanceOf(
  _type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  return {
    name: 'instanceOf',
    value: printValue(argumentPath),
  };
}

const simplePropTypes = [
  'array',
  'bool',
  'func',
  'number',
  'object',
  'string',
  'any',
  'element',
  'node',
  'symbol',
  'elementType',
] as const;

function isSimplePropType(
  name: string,
): name is (typeof simplePropTypes)[number] {
  return simplePropTypes.includes(name as (typeof simplePropTypes)[number]);
}

type PropTypeHandler = (
  type: PropTypeDescriptor,
  argumentPath: NodePath,
) => PropTypeDescriptor;

const propTypes = new Map<
  string,
  (argumentPath: NodePath | undefined) => PropTypeDescriptor
>([
  ['oneOf', callPropTypeHandler.bind(null, 'enum', getPropTypeOneOf)],
  ['oneOfType', callPropTypeHandler.bind(null, 'union', getPropTypeOneOfType)],
  [
    'instanceOf',
    callPropTypeHandler.bind(null, 'instanceOf', getPropTypeInstanceOf),
  ],
  ['arrayOf', callPropTypeHandler.bind(null, 'arrayOf', getPropTypeArrayOf)],
  ['objectOf', callPropTypeHandler.bind(null, 'objectOf', getPropTypeObjectOf)],
  ['shape', callPropTypeHandler.bind(null, 'shape', getPropTypeShapish)],
  ['exact', callPropTypeHandler.bind(null, 'exact', getPropTypeShapish)],
]);

function callPropTypeHandler(
  name: PropTypeDescriptor['name'],
  handler: PropTypeHandler,
  argumentPath: NodePath | undefined,
): PropTypeDescriptor {
  let type: PropTypeDescriptor = { name };

  if (argumentPath) {
    type = handler(type, argumentPath);
  }

  if (!type.value) {
    // If there is no argument then leave the value an empty string
    type.value = argumentPath ? printValue(argumentPath) : '';
    type.computed = true;
  }

  return type;
}

/**
 * Tries to identify the prop type by inspecting the path for known
 * prop type names. This method doesn't check whether the found type is actually
 * from React.PropTypes. It simply assumes that a match has the same meaning
 * as the React.PropTypes one.
 *
 * If there is no match, "custom" is returned.
 */
export default function getPropType(path: NodePath): PropTypeDescriptor {
  let descriptor: PropTypeDescriptor | null = null;

  getMembers(path, true).some((member) => {
    const memberPath = member.path;
    let name: string | null = null;

    if (memberPath.isStringLiteral()) {
      name = memberPath.node.value;
    } else if (memberPath.isIdentifier() && !member.computed) {
      name = memberPath.node.name;
    }
    if (name) {
      if (isSimplePropType(name)) {
        descriptor = { name };

        return true;
      }
      const propTypeHandler = propTypes.get(name);

      if (propTypeHandler) {
        descriptor = propTypeHandler(member.argumentPaths[0]);

        return true;
      }
    }

    return;
  });

  if (descriptor) {
    return descriptor;
  }

  if (path.isIdentifier() && isSimplePropType(path.node.name)) {
    return { name: path.node.name };
  }

  if (path.isCallExpression()) {
    const callee = path.get('callee');

    if (callee.isIdentifier()) {
      const propTypeHandler = propTypes.get(callee.node.name);

      if (propTypeHandler) {
        return propTypeHandler(path.get('arguments')[0]);
      }
    }
  }

  return { name: 'custom', raw: printValue(path) };
}
