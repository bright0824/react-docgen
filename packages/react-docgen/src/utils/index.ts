import * as docblock from './docblock';
import * as expressionTo from './expressionTo';
import * as flowUtilityTypes from './flowUtilityTypes';
import * as traverse from './traverse';

export { docblock, expressionTo, flowUtilityTypes, traverse };

export { default as getClassMemberValuePath } from './getClassMemberValuePath';
export { default as getFlowType } from './getFlowType';
export { default as getMemberExpressionRoot } from './getMemberExpressionRoot';
export { default as getMemberExpressionValuePath } from './getMemberExpressionValuePath';
export { default as getMembers } from './getMembers';
export {
  default as getMemberValuePath,
  isSupportedDefinitionType,
} from './getMemberValuePath';
export { default as getMethodDocumentation } from './getMethodDocumentation';
export type { MethodNodePath } from './getMethodDocumentation';
export { default as getNameOrValue } from './getNameOrValue';
export { default as getParameterName } from './getParameterName';
export { default as getPropertyName, COMPUTED_PREFIX } from './getPropertyName';
export { default as getPropertyValuePath } from './getPropertyValuePath';
export { default as getPropType } from './getPropType';
export { default as getTSType } from './getTSType';
export { default as getTypeAnnotation } from './getTypeAnnotation';
export {
  default as getTypeFromReactComponent,
  applyToTypeProperties,
} from './getTypeFromReactComponent';
export { default as getTypeIdentifier } from './getTypeIdentifier';
export { default as getTypeParameters } from './getTypeParameters';
export type { TypeParameters } from './getTypeParameters';
export { default as isDestructuringAssignment } from './isDestructuringAssignment';
export { default as isExportsOrModuleAssignment } from './isExportsOrModuleAssignment';
export { default as isReactBuiltinCall } from './isReactBuiltinCall';
export { default as isReactChildrenElementCall } from './isReactChildrenElementCall';
export { default as isReactCloneElementCall } from './isReactCloneElementCall';
export { default as isReactComponentClass } from './isReactComponentClass';
export { default as isReactComponentMethod } from './isReactComponentMethod';
export { default as isReactCreateClassCall } from './isReactCreateClassCall';
export { default as isReactCreateElementCall } from './isReactCreateElementCall';
export { default as isReactForwardRefCall } from './isReactForwardRefCall';
export { default as isReactModuleName } from './isReactModuleName';
export { default as isRequiredPropType } from './isRequiredPropType';
export { default as isStatelessComponent } from './isStatelessComponent';
export { default as isUnreachableFlowType } from './isUnreachableFlowType';
export { default as match } from './match';
export { default as normalizeClassDefinition } from './normalizeClassDefinition';
export { default as parseJsDoc } from './parseJsDoc';
export { default as postProcessDocumentation } from './postProcessDocumentation';
export { default as printValue } from './printValue';
export { default as resolveExportDeclaration } from './resolveExportDeclaration';
export { default as resolveFunctionDefinitionToReturnValue } from './resolveFunctionDefinitionToReturnValue';
export { default as resolveGenericTypeAnnotation } from './resolveGenericTypeAnnotation';
export { default as resolveHOC } from './resolveHOC';
export { default as resolveObjectPatternPropertyToValue } from './resolveObjectPatternPropertyToValue';
export {
  default as resolveObjectKeysToArray,
  resolveObjectToNameArray,
} from './resolveObjectKeysToArray';
export { default as resolveObjectValuesToArray } from './resolveObjectValuesToArray';
export { default as resolveToModule } from './resolveToModule';
export { default as resolveToValue } from './resolveToValue';
export { default as setPropDescription } from './setPropDescription';
