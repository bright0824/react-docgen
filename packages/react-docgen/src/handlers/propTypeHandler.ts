import getPropType from '../utils/getPropType.js';
import getPropertyName from '../utils/getPropertyName.js';
import getMemberValuePath from '../utils/getMemberValuePath.js';
import isReactModuleName from '../utils/isReactModuleName.js';
import isRequiredPropType from '../utils/isRequiredPropType.js';
import printValue from '../utils/printValue.js';
import resolveToModule from '../utils/resolveToModule.js';
import resolveToValue from '../utils/resolveToValue.js';
import type Documentation from '../Documentation.js';
import type { PropDescriptor, PropTypeDescriptor } from '../Documentation.js';
import type { NodePath } from '@babel/traverse';
import type { Node } from '@babel/types';
import type { Handler } from './index.js';
import type { ComponentNode } from '../resolver/index.js';

function isPropTypesExpression(path: NodePath): boolean {
  const moduleName = resolveToModule(path);

  if (moduleName) {
    return isReactModuleName(moduleName) || moduleName === 'ReactPropTypes';
  }

  return false;
}

function amendPropTypes(
  getDescriptor: (propName: string) => PropDescriptor,
  path: NodePath,
): void {
  if (!path.isObjectExpression()) {
    return;
  }

  path.get('properties').forEach(propertyPath => {
    if (propertyPath.isObjectProperty()) {
      const propName = getPropertyName(propertyPath);

      if (!propName) return;

      const propDescriptor = getDescriptor(propName);
      const valuePath = resolveToValue(propertyPath.get('value'));
      const type: PropTypeDescriptor = isPropTypesExpression(valuePath)
        ? getPropType(valuePath)
        : { name: 'custom', raw: printValue(valuePath) };

      if (type) {
        propDescriptor.type = type;
        propDescriptor.required =
          type.name !== 'custom' && isRequiredPropType(valuePath);
      }
    }
    if (propertyPath.isSpreadElement()) {
      const resolvedValuePath = resolveToValue(propertyPath.get('argument'));

      if (resolvedValuePath.isObjectExpression()) {
        // normal object literal
        amendPropTypes(getDescriptor, resolvedValuePath);
      }
    }
  });
}

function getPropTypeHandler(propName: string): Handler {
  return function (
    documentation: Documentation,
    componentDefinition: NodePath<ComponentNode>,
  ): void {
    let propTypesPath: NodePath<Node> | null = getMemberValuePath(
      componentDefinition,
      propName,
    );

    if (!propTypesPath) {
      return;
    }
    propTypesPath = resolveToValue(propTypesPath);
    if (!propTypesPath) {
      return;
    }
    let getDescriptor;

    switch (propName) {
      case 'childContextTypes':
        getDescriptor = documentation.getChildContextDescriptor;
        break;
      case 'contextTypes':
        getDescriptor = documentation.getContextDescriptor;
        break;
      default:
        getDescriptor = documentation.getPropDescriptor;
    }
    amendPropTypes(getDescriptor.bind(documentation), propTypesPath);
  };
}

export const propTypeHandler = getPropTypeHandler('propTypes');
export const contextTypeHandler = getPropTypeHandler('contextTypes');
export const childContextTypeHandler = getPropTypeHandler('childContextTypes');
