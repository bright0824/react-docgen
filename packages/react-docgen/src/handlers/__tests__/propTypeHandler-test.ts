import { parse, makeMockImporter, noopImporter } from '../../../tests/utils';
import Documentation from '../../Documentation';
import type DocumentationMock from '../../__mocks__/Documentation';
import { propTypeHandler } from '../propTypeHandler.js';
import getPropType from '../../utils/getPropType';
import type { NodePath } from '@babel/traverse';
import type { Importer } from '../../importer';
import type {
  ArrowFunctionExpression,
  ClassDeclaration,
  FunctionDeclaration,
  ObjectExpression,
} from '@babel/types';
import type { ComponentNode } from '../../resolver';
import { beforeEach, describe, expect, test, vi } from 'vitest';

const getPropTypeMock = vi.mocked(getPropType);

vi.mock('../../Documentation.js');
vi.mock('../../utils/getPropType.js', () => ({ default: vi.fn(() => ({})) }));

describe('propTypeHandler', () => {
  let documentation: Documentation & DocumentationMock;

  beforeEach(() => {
    documentation = new Documentation() as Documentation & DocumentationMock;
  });

  const mockImporter = makeMockImporter({
    props: (stmtLast) =>
      stmtLast(`
      import { PropTypes } from 'react';
      export default {bar: PropTypes.bool};
    `).get('declaration'),

    simpleProp: (stmtLast) =>
      stmtLast(`
      import { PropTypes } from 'react';
      export default PropTypes.array.isRequired;
    `).get('declaration'),

    complexProp: (stmtLast) =>
      stmtLast(`
      var prop = PropTypes.oneOfType([PropTypes.number, PropTypes.bool]).isRequired;
      import { PropTypes } from 'react';
      export default prop;
    `).get('declaration'),

    foo: (stmtLast) =>
      stmtLast(`
        import { PropTypes } from 'react';
        export default PropTypes.bool;
    `).get('declaration'),

    bar: (stmtLast) =>
      stmtLast(`
        import { PropTypes } from 'react';
        export default PropTypes.bool;
    `).get('declaration'),

    baz: (stmtLast) =>
      stmtLast(`
        import { PropTypes as OtherPropTypes } from 'react';
        export default OtherPropTypes.bool;
    `).get('declaration'),
  });

  function template(src: string) {
    return `
      ${src}
      var React = require('React');
      var PropTypes = React.PropTypes;
      var {PropTypes: OtherPropTypes} = React;
    `;
  }

  function testPropTypes(
    getSrc: (src: string) => string,
    parseSrc: (src: string, importer?: Importer) => NodePath<ComponentNode>,
  ) {
    test('passes the correct argument to getPropType', () => {
      const propTypesSrc = `{
          foo: PropTypes.bool,
          abc: PropTypes.xyz,
        }`;
      const definition = parseSrc(getSrc(propTypesSrc));

      propTypeHandler(documentation, definition);

      expect(getPropTypeMock.mock.calls[0][0]).toMatchSnapshot();
      expect(getPropTypeMock.mock.calls[1][0]).toMatchSnapshot();
    });

    test('finds definitions via React.PropTypes', () => {
      const definition = parseSrc(
        getSrc(
          `{
          foo: PropTypes.bool,
          bar: require("react").PropTypes.bool,
          baz: OtherPropTypes.bool,
        }`,
        ),
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toEqual({
        foo: {
          type: {},
          required: false,
        },
        bar: {
          type: {},
          required: false,
        },
        baz: {
          type: {},
          required: false,
        },
      });
    });

    test('finds definitions via the ReactPropTypes module', () => {
      const definition = parseSrc(
        getSrc(
          `{
          foo: require("ReactPropTypes").bool,
        }`,
        ),
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toEqual({
        foo: {
          type: {},
          required: false,
        },
      });
    });

    test('detects whether a prop is required', () => {
      const definition = parseSrc(
        getSrc(
          `{
          simple_prop: PropTypes.array.isRequired,
          complex_prop:
            PropTypes.oneOfType([PropTypes.number, PropTypes.bool]).isRequired,
        }`,
        ),
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toEqual({
        simple_prop: {
          type: {},
          required: true,
        },
        complex_prop: {
          type: {},
          required: true,
        },
      });
    });

    test('handles computed properties', () => {
      const definition = parseSrc(
        getSrc(
          `{
          [foo]: PropTypes.array.isRequired,
          complex_prop:
            PropTypes.oneOfType([PropTypes.number, PropTypes.bool]).isRequired,
        }`,
        ),
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('ignores complex computed properties', () => {
      const definition = parseSrc(
        getSrc(
          `{
          [() => {}]: PropTypes.array.isRequired,
          complex_prop:
            PropTypes.oneOfType([PropTypes.number, PropTypes.bool]).isRequired,
        }`,
        ),
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('only considers definitions from React or ReactPropTypes', () => {
      const definition = parseSrc(
        getSrc(
          `{
          custom_propA: PropTypes.bool,
          custom_propB: Prop.bool.isRequired,
        }`,
        ),
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toEqual({
        custom_propA: {
          type: {},
          required: false,
        },
        custom_propB: {
          type: {
            name: 'custom',
            raw: 'Prop.bool.isRequired',
          },
          required: false,
        },
      });
    });

    test('resolves variables', () => {
      const definitionSrc = getSrc('props');
      const definition = parseSrc(`
        ${definitionSrc}
        var props = {bar: PropTypes.bool};
      `);

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toEqual({
        bar: {
          type: {},
          required: false,
        },
      });
    });

    test('resolves imported variables', () => {
      const definitionSrc = getSrc('props');
      const definition = parseSrc(
        `
        ${definitionSrc}
        import props from 'props';
      `,
        mockImporter,
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toEqual({
        bar: {
          type: {},
          required: false,
        },
      });
    });

    test('can resolve individual imported variables assigned to props', () => {
      const definitionSrc = getSrc(`{
        foo: foo,
        [bar]: bar,
        baz: baz,
        simple_prop: simpleProp,
        complex_prop: complexProp,
      }`);
      const definition = parseSrc(
        `
        ${definitionSrc}
        import foo from 'foo';
        import bar from 'bar';
        import baz from 'baz';
        import simpleProp from 'simpleProp';
        import complexProp from 'complexProp';
      `,
        mockImporter,
      );

      propTypeHandler(documentation, definition);
      expect(documentation.descriptors).toMatchSnapshot();
    });
  }

  describe('React.createClass', () => {
    testPropTypes(
      (propTypesSrc) => template(`({propTypes: ${propTypesSrc}})`),
      (src, importer = noopImporter) =>
        parse
          .statement(src, importer)
          .get('expression') as NodePath<ObjectExpression>,
    );
  });

  describe('class definition', () => {
    describe('class property', () => {
      testPropTypes(
        (propTypesSrc) =>
          template(`
          class Component {
            static propTypes = ${propTypesSrc};
          }
        `),
        (src, importer = noopImporter) => parse.statement(src, importer),
      );
    });

    describe('static getter', () => {
      testPropTypes(
        (propTypesSrc) =>
          template(`
          class Component {
            static get propTypes() {
              return ${propTypesSrc};
            }
          }
        `),
        (src, importer = noopImporter) => parse.statement(src, importer),
      );
    });
  });

  describe('stateless component', () => {
    testPropTypes(
      (propTypesSrc) =>
        template(`
        var Component = (props) => <div />;
        Component.propTypes = ${propTypesSrc};
      `),
      (src, importer = noopImporter) => parse.statement(src, importer),
    );
  });

  describe('does not error if propTypes cannot be found', () => {
    test('ObjectExpression', () => {
      const definition = parse.expression<ObjectExpression>('{fooBar: 42}');

      expect(() => propTypeHandler(documentation, definition)).not.toThrow();
    });

    test('ClassDeclaration', () => {
      const definition = parse.statement<ClassDeclaration>('class Foo {}');

      expect(() => propTypeHandler(documentation, definition)).not.toThrow();
    });

    test('FunctionDeclaration', () => {
      const definition =
        parse.statement<FunctionDeclaration>('function Foo() {}');

      expect(() => propTypeHandler(documentation, definition)).not.toThrow();
    });

    test('ArrowFunctionExpression', () => {
      const definition = parse.expression<ArrowFunctionExpression>('() => {}');

      expect(() => propTypeHandler(documentation, definition)).not.toThrow();
    });
  });

  // This case is handled by propTypeCompositionHandler
  test('does not error if propTypes is a member expression', () => {
    const definition = parse.expression<ObjectExpression>(
      '{propTypes: Foo.propTypes}',
    );

    expect(() => propTypeHandler(documentation, definition)).not.toThrow();
  });
});
