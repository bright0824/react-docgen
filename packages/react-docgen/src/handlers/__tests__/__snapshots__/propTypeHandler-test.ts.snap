// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propTypeHandler React.createClass can resolve individual imported variables assigned to props 1`] = `
{
  "@computed#bar": {
    "required": false,
    "type": {},
  },
  "baz": {
    "required": false,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
  "foo": {
    "required": false,
    "type": {},
  },
  "simple_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler React.createClass handles computed properties 1`] = `
{
  "@computed#foo": {
    "required": true,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler React.createClass ignores complex computed properties 1`] = `
{
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler class definition class property can resolve individual imported variables assigned to props 1`] = `
{
  "@computed#bar": {
    "required": false,
    "type": {},
  },
  "baz": {
    "required": false,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
  "foo": {
    "required": false,
    "type": {},
  },
  "simple_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler class definition class property handles computed properties 1`] = `
{
  "@computed#foo": {
    "required": true,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler class definition class property ignores complex computed properties 1`] = `
{
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler class definition static getter can resolve individual imported variables assigned to props 1`] = `
{
  "@computed#bar": {
    "required": false,
    "type": {},
  },
  "baz": {
    "required": false,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
  "foo": {
    "required": false,
    "type": {},
  },
  "simple_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler class definition static getter handles computed properties 1`] = `
{
  "@computed#foo": {
    "required": true,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler class definition static getter ignores complex computed properties 1`] = `
{
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler stateless component can resolve individual imported variables assigned to props 1`] = `
{
  "@computed#bar": {
    "required": false,
    "type": {},
  },
  "baz": {
    "required": false,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
  "foo": {
    "required": false,
    "type": {},
  },
  "simple_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler stateless component handles computed properties 1`] = `
{
  "@computed#foo": {
    "required": true,
    "type": {},
  },
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;

exports[`propTypeHandler stateless component ignores complex computed properties 1`] = `
{
  "complex_prop": {
    "required": true,
    "type": {},
  },
}
`;
